src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java

```diff
@@ -609,11 +628,26 @@ else if (assignDepth >= 1) {
     private static boolean unnecessaryParenAroundOperators(DetailAST ast) {
         final int type = ast.getType();
         final int parentType = ast.getParent().getType();
-        final boolean isConditional = TokenUtil.isOfType(type, CONDITIONALS_AND_RELATIONAL);
-        boolean result = TokenUtil.isOfType(parentType, CONDITIONALS_AND_RELATIONAL);
-        if (isConditional) {
-            if (type == TokenTypes.LOR) {
-                result = result && !TokenUtil.isOfType(parentType, TokenTypes.LAND);
+        final boolean isConditionalOrBitwiseBinaryOperator =
+                isConditionalOrBitwiseBinary(type);
+        boolean result = isConditionalOrBitwiseBinary(parentType);
+        if (isConditionalOrBitwiseBinaryOperator) {
```

Comment:
Are we forced to couple checking of conditional and binary operators? Can we do something like:
```
if (isConditionalOperator) {
...
}
else if (isBitWiseBinaryOperator) {
...
}

```

Reply:
we  can make implemenation like 
```
    private static boolean unnecessaryParenAroundOperators(DetailAST ast) {
        final int type = ast.getType();
        final int parentType = ast.getParent().getType();
        final boolean isConditional = TokenUtil.isOfType(type, CONDITIONALS_AND_RELATIONAL);
        final boolean isBitwise = TokenUtil.isOfType(type, BITWISE_BINARY_OPERATORS);
        boolean result = isConditionalOrBitwiseBinary(parentType);
        if (isConditional) {
            if (type == TokenTypes.LOR) {
                result = result(parentType, TokenTypes.LAND)
                        && result(parentType, BITWISE_BINARY_OPERATORS);
            }
            if (type == TokenTypes.LAND) {
                result = result(parentType, BITWISE_BINARY_OPERATORS);
            }
            result = result && !TokenUtil.isOfType(parentType, TokenTypes.EQUAL,
                TokenTypes.NOT_EQUAL);
        }
        else if (isBitwise) {
             if (type == TokenTypes.BOR) {
                result = result(parentType, TokenTypes.BAND)
                        && result(parentType, TokenTypes.BXOR);
            }
            if (type == TokenTypes.BXOR) {
                result = result(parentType, TokenTypes.BAND);
            }
            result = result && !TokenUtil.isOfType(parentType, TokenTypes.EQUAL,
                TokenTypes.NOT_EQUAL);
        }
        else {
            result = result && TokenUtil.isOfType(type, UNARY_AND_POSTFIX);
        }
        return result;
    }
```

but  cyclomatic complexity is 12  here but allowed is 10. I have tried a lot but  
The way i implement(switch case) is the only way which i found

Reply:
Please do:
```
if (isConditionalOperator) {
    result = checkConditionalOperator(...);
}
else if (isBitWiseBinaryOperator) {
    result = checkBitWiseBinaryOperator(...);
}
else {
...
}

```

And remove other helper methods (like `result()` so that we can clearly see logic involved in each check.

Reply:
Why are we required to consider parent at all in this method for conditional/bitwise operators? We should handle all logic for these in `check...` method.

Reply:
I am sorry :) but i think i am missing something from https://github.com/checkstyle/checkstyle/pull/12703#discussion_r1148748480 to understand what changes you suggesting.  

i think you want me to remove `TokenUtil.isOfType(parentType, CONDITIONALS_AND_RELATIONAL || TokenUtil.isOfType(parentType, BITWISE_BINARY_OPERATORS`  am i right?

I have tried to not try to make any new logic for the token i have only enhance the current code to make not more confusion. the logic is same

Reply:
Current code (in master branch) in this method is not easy to understand, this is why we exceeded cyclomatic complexity with your changes. We need to push logic to check operators out to helper methods such that method reads like:
```
final boolean result;
if (isConditionalOperator) {
    result = checkConditionalOperator(...);
}
else if (isBitWiseBinaryOperator) {
    result = checkBitWiseBinaryOperator(...);
}
else {
// handle unary and postfix here
...
}
```

Instead of doing what I suggested, your changes to the method have some repetitive logic (checking of parents for bitwise and conditional operators), and couple checking of bitwise operators and conditional operators.

I am open to other ways to handle update to this method, but it must be easy to understand and reason about, which neither the existing code, nor the most recent changes in this PR are.

Reply:
please check again i have update the branch

Reply:
done